File: ios/App/Sources/ChapbookAppModel/Catalogs.swift
import Foundation

/// A catalog the reader added: where it is and what it called itself.
public struct SavedCatalog: Codable, Hashable, Sendable, Identifiable {
    public let id: String
    public var title: String
    public let url: String

    public init(id: String, title: String, url: String) {
        self.id = id
        self.title = title
        self.url = url
    }
}

/// The catalogs the reader has added, in the order they were added.
///
/// Nothing here is a secret — a catalog's credential lives in
/// `Credentials` under its origin — so defaults hold the list. The
/// engine has no notion of a saved catalog; the shelf only knows the
/// books that came from one.
@MainActor
public final class Catalogs: ObservableObject {
    private let defaults: UserDefaults
    private static let key = "catalogs"

    @Published public private(set) var all: [SavedCatalog]

    public init(defaults: UserDefaults) {
        self.defaults = defaults
        all = Self.load(defaults)
    }

    public func get(_ id: String) -> SavedCatalog? {
        all.first { $0.id == id }
    }

    /// Add a catalog; `title` may be blank until its feed says what it is
    /// called.
    @discardableResult
    public func add(url: String, title: String = "") -> SavedCatalog {
        let catalog = SavedCatalog(
            id: UUID().uuidString,
            title: title.trimmingCharacters(in: .whitespaces),
            url: url.trimmingCharacters(in: .whitespaces))
        save(all + [catalog])
        return catalog
    }

    public func rename(_ id: String, to title: String) {
        save(all.map { $0.id == id ? SavedCatalog(id: $0.id, title: title, url: $0.url) : $0 })
    }

    public func remove(_ id: String) {
        save(all.filter { $0.id != id })
    }

    private func save(_ list: [SavedCatalog]) {
        if let data = try? JSONEncoder().encode(list) {
            defaults.set(data, forKey: Self.key)
        }
        all = list
    }

    private static func load(_ defaults: UserDefaults) -> [SavedCatalog] {
        guard let data = defaults.data(forKey: key) else { return [] }
        return (try? JSONDecoder().decode([SavedCatalog].self, from: data)) ?? []
    }
}

File: android/app/src/main/kotlin/com/ophymx/chapbook/app/model/Catalogs.kt
package com.ophymx.chapbook.app.model

import com.ophymx.chapbook.App
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow

/** A catalog the reader added: where it is and what it called itself. */
data class SavedCatalog(val id: String, val title: String, val url: String)

/**
 * The catalogs the reader has added, in the order they were added, as
 * the engine keeps them beside the shelf. A catalog's credential lives
 * in [Credentials] under its origin, never here.
 *
 * The ids are the library's rows, carried as strings because that is
 * what a navigation route holds. Each call is one small query on the
 * shelf's thread, waited for: an add has to answer with its row so the
 * screen can open it, and the wait is shorter than a frame.
 */
class Catalogs(private val shelf: Shelf) {
    private val _all = MutableStateFlow<List<SavedCatalog>>(emptyList())
    val all: StateFlow<List<SavedCatalog>> = _all

    init {
        refresh()
    }

    fun get(id: String): SavedCatalog? = _all.value.firstOrNull { it.id == id }

    /** Add a catalog; [title] may be blank until its feed says what it is called. */
    fun add(url: String, title: String = ""): SavedCatalog {
        val added = checkNotNull(shelf.blockingApp { addCatalog(url, title) }) { "the catalog could not be added; see logcat" }
        refresh()
        return added.saved()
    }

    fun rename(id: String, title: String) {
        shelf.blockingApp { renameCatalog(id.toLong(), title) }
        refresh()
    }

    fun remove(id: String) {
        shelf.blockingApp { removeCatalog(id.toLong()) }
        refresh()
    }

    private fun refresh() {
        _all.value = shelf.blockingApp { catalogs().map { it.saved() } }
    }

    private fun App.CatalogRecord.saved() = SavedCatalog(id.toString(), title, url)
}
